### DISCUSSION TITLE
Longer term thoughts

### DISCUSSION THREAD

**Discussion by rob-bygrave** [ebean-orm/ebean#3112] [General]
Ebean is getting close to 17 years old now and a few things have me thinking about what the next 10 odd years might look like.

In case you missed it Oracle announced that the Graal Enterprise features are going to be free. To me this is another pointer that the future is about smaller, simpler, faster, lighter and probably "more native compilation"- that native-image is going to come to  larger adoption sooner than I previously thought.

I'm conscience that ebea-core is a ~2Mb jar (and I also see another major ORM has it's core jar coming in at ~ 11Mb). The JPA spec was pretty much completely dormant for a long long time now but now has popped back to life. To some extent that is good but it perhaps is a reminder that Ebean naturally took on JPA Mapping support when some of it was in my opinion poor and some of it now looks dated. I'm especially aware that I personally really don't like `@ElementCollection` and I don't like `@Inheritance` and I have never chosen to use either - ever !!!.

I'm also aware that at this point we are very unlikely to see more assistance in the form of contributing to docs, examples, code or sponsors that fund more time for docs, examples, code etc. We need to be aware of what Ebean can take on in terms of features and complexity.

With these background thoughts I have in mind to do a little bit of research in terms of how much simplicity we can get from removing some features like for example Named queries (no more antlr4 dependency etc).

Cheers, Rob.

**Comment by jnehlmeier** [comment]
Keep in mind that if you remove something you also make it more work to migrate to Ebean.

Given that GraalVM native-image only compiles what is reachable + what it is being told to treat reachable, the size of the library doesn't have a huge impact.

Personally it would be way better to give documentation a stronger focus. Without proper documentation, adoption is harder. Without knowing Ebean concepts on a technical level and without proper JavaDoc on internal classes contribution is also harder.

**Comment by rbygrave** [comment]
> Given that GraalVM ...

Yes, unused code removal is good there. The size of the lib will still impact compile time.  Hard to say that the extra complexity for features like elementcollection and inheritance can be optimized out though - not sure.

> documentation a stronger focus

Yes. 

> if you remove something you also make it more work to migrate

Yes. Feature removal is a serious call/issue. It impacts existing users. We'd do it for an important reason (like avoiding Rob burnout).

**Comment by rPraml** [comment]
>  I'm especially aware that I personally really don't like `@ElementCollection` and I don't like `@Inheritance` and I have never chosen to use either - ever !!!.

FYI: We do not use `@ElementCollection` in our application.
But we use `@Inheritance`.
We tend to remove `@Inheritance` on our DB-Models as it makes problems on some places.
We use `@Inheritance` also a lot in conjunction with `@DbJson` + `@DocStore` where we still need it. (It would take some effort, but would be possible to use jackson for deserialization here)

**Reply by rbygrave** [reply]
> @DocStore

Just to be clear, `@DocStore` is the same as _"Elastic Search" support_ which is on the list for removal in 15.x.

**Reply by rPraml** [reply]
We use `@DocStore` often in conjunction with `@DbJson`

Here at FOCONIS, `@DocStore` means, that the EntityBean is stored in JSON format in an other `@Entity` - There are also many tests provided that do this.

Maybe we misused this annotation. The above should be possible with `@Entity` also (but we do not want a table)
Question may appear:
Q: Why do we need `@DocStore` at all and don't use plain Jackson/etc to serialize the bean.
A: We want the enhancement, we want to traverse the graph with the ebean descriptors and also want the old/new values. We also want to refer real entites from `@DocStore`

Maybe, the annotation can be renamed to `@JsonEntity` - which means that this is an `@Entity` but has no physical table in the database.

**Reply by rob-bygrave** [reply]
Ah ok, that is an interesting use. Good to know.  A little bit like a `@View` based entity or maybe a special embeddable. Is it only used as an "embedded document" inside an actual `@Entity`? 

>  Maybe, the annotation can be renamed to @JsonEntity - which means that this is an @Entity but has no physical table in the database.

Yeah. Ultimately we don't want it confused with the ElasticSearch feature and that the name + documentation + tests makes it clear and obvious how it is expected to be used. [I don't recall tests for this, do we have tests for this?]

**Comment by rbygrave** [comment]
FYI: The full list of features that I'm thinking about for this experiment

## Going to keep
- findFutureIds() findFutureList() ... (was initially removed in 15.0.0-RC and restored in 15.0.1)
- Support for modelling `@OneToMany` as Maps
- Support for modelling `@OneToMany` as Sets
Some good enough reasons to keep these features as discussed above. Ideally better docs here for use of Set with hashcode/equals.

## Changes in `main` to support this

- [x] add filterManyRaw to use raw() rather than parsing expression (to support removal of query language parsing) 

## Probably not too controversial

- [x] Remove Expression Query Language [parsing] & Named queries
       ...  Removes ANTLR  dependency
       ... removes named queries
- [x] Remove External mapping (XML mapping, named queries - suspected to be rarely used these days, currently optional)
       ... removes named queries
- [x] Remove Read Auditing

## Controversial

- [x] Remove `@Draftable`
- [x] Remove @Inheritance (Existing code would require to flatten model onto root bean, *could be painful)
- [x] Remove ElasticSearch support

## Features discussed but ultimately _not_ removed or changed in 15.x
- Replacing [SQL] @Formula, with a [Logical property based] @Formula2 ?? *could be painful
- Entities as JSON support
- Support for mapping java.util.Date, Calendar (Use newer java time types)
- @ElementCollection (promote use DB JSON columns going forward, *could be painful)
- L2 sharableBean (an internal detail of L2 cache)

**Comment by nPraml** [comment]
> * [ ]  Remove support for modelling `@OneToMany` as Maps (support List and Set only)
> * [ ]  Remove support for modelling `@OneToMany` as Sets (support List only)
> 

@rob-bygrave can you tell me why you don't want to support Sets in the future?
We have some models that have Set fields on purpose. This automatically prevents having two identical elements in the collection.

**Reply by rbygrave** [reply]
> why you don't want to support Sets in the future?

Removing support for Set is saying that there are NO known good use cases where we recommend its use - NONE.

For Sets to be provide some value then the need some hashCode/equals implementation that effectively de-duplicates elements. The question becomes, is that _ever_ a good idea. Is it ever a good idea to have rows in the database mapped into a Set and be implicitly de-duplicated in-memory on the application side _always_ and _automatically_, as opposed to explicit de-duplication as and when needed.

The only cases that I had hit have been better served by using `findMap()` or explicit application code.

Said differently, I know of NO known good use cases for Set but I know of several ways that people can mess it up. 

Just in the last week, we see #3173 and the question is, is that a good use of Set or actually a case where it should instead have been modelled as List and for the de-duplicate to be explicit? At this stage, I suspect when I see the failing test case for #3173 it will re-enforce the idea that Set is not good.  


> This automatically prevents having two identical elements in the collection.

I'd be interested to see what this looks like in practice and why it's thought of as a good idea, why was Set was chosen, why was it chosen to de-dup via the model rather than explicitly etc.

**Reply by jnehlmeier** [reply]
I looked through some code and `Set` is usually used with enums (`EnumSet` in constructor initialization, corresponding methods are `Set` based). There are also some occurrences of `Set` in combination with a `SortedSet` coming from application code but this doesn't make too much sense as it looses order when read from DB. Also some `@OneToMany` fields are `Collection` probably because IDE suggested it as no `List` specific method was ever used. Overall the vast majority of `@OneToMany` fields are `List`.

For `Map` it is usually used to speed up read access. The key is usually some application managed identifier, e.g. a `String` or `Enum`. `Map` is used to avoid searching all the time and and removing `Map` support would force introducing a `transient Map` field for reads. So it would make code look more ugly. As an example consider a questionnaire with questions and each question has a unique identifier. Answers are stored in a questionnaire result using that unique identifier. If you now have some application logic on answers it usually operates on these unique identifiers coming from the questionnaire model. So a `Map` for read access is desired.

**Reply by rPraml** [reply]
> I'd be interested to see what this looks like in practice and why it's thought of as a good idea, why was Set was chosen, why was it chosen to de-dup via the model rather than explicitly etc.

We often have code, that adds something, e.g with should be deduped. Mostly that are user objects, but here is a good example in a test case:
```java
class Role {
    @ManyToMany
    Set<Permission> permissions;
}
```

If I want to add two permissions, I just call `role.getPermissions().add(p1)`, `role.getPermissions().add(p2)`
If I have a list here, I have to convert it in a set first:

```java
Set<Permission> tmp = new HashSet(role.getPermissions());
tmp.add(p1);
tmp.add(p2)
role.getPermissions().clear();
role.getPermissions().addAll(tmp);
```

Of course, I can also implement it with List.contains, but I always must iterate through all elements of the list. 
List.contains has complexity of (`O(n)`) instead of the `O(1)` performance of Set.add

The problem is, if adding `p1` and `p2` is on different code places, I have to build the temp set multiple times.

It would be a good option, if we can implement some `shadow` fields:

```java
class Role {
    @ManyToMany
    List<Permission> _permissions;

    @Transient
    Set<Permission> permissions;

    Set<Permission> getPermissions() {
       if (permissions == null) {
           permissions = new LinkedHashSet(_permissions); // or use a list, that invokes a callback on modification
       }
       return permissions;
    }

   void methodMustBeCalledBeforePersistEvenIfBeanIsNotDirty() {
       if (permissions != null) {
         _permissions.clear();
         _permissions.addAll(permissions);
      }
   }
}
```

This can be done for maps also, but ebean needs some mechanism to call a method, that writes back the state of the shadow field.
Unfortunately, this targets also dirty recognition, but would be doable.

**Reply by rob-bygrave** [reply]
Well, I think I'm wanting to see more accurately **_HOW_** Set is being used and take stock of the tests that we have. What exactly does `Permission` look like and especially the hashCode/equals implementations.

We could decide to keep support for Set if we are clear on how it can be used properly [perhaps we give **_Excellent_** guidance on hashCode/equals implementations].

**Comment by jnehlmeier** [comment]
> Remove Expression Query Language [parsing] & Named queries
... Removes ANTLR dependency
... removes named queries
 Remove External mapping (XML mapping, named queries - suspected to be rarely used these days, currently optional)
... removes named queries

Never used this even in JPA. 

> Remove findFutureIds() findFutureList() ...

No opinion, I never asked an ORM before to query something in a background thread. Usually the background thread has been created beforehand in the application. 

> Remove L2 sharableBean (an internal detail of L2 cache)

Always good to hide internal stuff which is then of course free to remove. 

> Remove Read Auditing

Seems useful for certain people that have stronger security requirements (heath care, banking, ..). Maybe it is possible to add some hooks to ebean core so that this feature could be extracted as an ebean extension?

Otherwise it has to be done in the DB itself but that might be more complicated as in the application. 

> Remove support for mapping java.util.Date, Calendar (Use newer java time types)

Would make migrating to ebean more complicated as long as JPA supports both. It is quite some work to refactor large applications.

> Remove support for modelling @OneToMany as Maps (support List and Set only)

Maps are useful to speed up read access. Removing maps would make everyones code more ugly by introducing a transient map for reads (would also increase memory a bit, given two data structures for the same thing)

> Remove support for modelling @OneToMany as Sets (support List only)

I sometimes use it for Enums with `EnumSet`.

**Reply by rbygrave** [reply]
What does a `@OneToMany` with an EnumSet look like?  Care to share an example? I would have though that would be a `@DbArray`.

**Reply by jnehlmeier** [reply]
My bad. I mixed something up. `EnumSet` is used with `@ElementCollection + @CollectionTable` and not `@OneToMany`. This is JPA code so no `@DbArray`. I also can't migrate to `@DbArray` easily because a lot of code still uses JPA. So database tables are shared by JPA entities and Ebean entities. Entities are duplicated because of weaving, which is really annoying but well. So there need to be some common ground between JPA and Ebean for a doable migration. Both libraries need to co-exist for a longer time and need to access the same db tables until all code paths have been rewritten in ebean. Big bang doesn't really work if databases are large and code base is large.

I did a quick search again and there are like 10 `Set<Entity>` fields that are either `@OneToMany` or `@ManyToMany`. Some of them because it makes sense for the Entity API to return a `Set`. Some of them to have an easy `set.addAll` call without checking for duplicates. There are also unique keys involved on DB level, especially on the `@ManyToMany` link tables.

**Reply by rob-bygrave** [reply]
> I also can't migrate to @DbArray easily because a lot of code still uses JPA.

Hmm, bit of a blow that.  ElementCollection might not be easy to deal with without built in support in the ORM itself. Hmm.

**Reply by rob-bygrave** [reply]
> findFutureIds() findFutureList() ... No opinion, I never asked an ORM before to query something in a background thread.

**_FYI:_** 17 odd years ago this started as the idea that ebean could _"Build object graphs in parallel"_ and this is effectively the remnants of that original idea. Today `findFutureCount()` is useful / used with `findPagedList()` [in parallel fetch the total row count]. As I see it, the original idea was basically a bad idea and I don't actually expect anyone to be using findFutureIds() or findFutureList() but they have never been removed - the last vestiges of that original idea that wasn't ultimately a good idea. Conceptually we could hide `findFutureCount()` so that it is only used as an implementation detail of `findPagedList()`.

**Comment by jnehlmeier** [comment]
As a side note: With Java 21 virtual threads it is probably also a good idea to look for synchronized blocks / object.wait() within ebean to avoid thread pinning. But I think many JDBC drivers will have trouble with this as well.

**Reply by rbygrave** [reply]
> a good idea to look for synchronized blocks / object.wait() within ebean to avoid thread pinning

Already done. I did that ages ago and so there is no use of synchronised in ebean.

> JDBC drivers will have trouble with this as well.

I'm also the guy that got the change into PostgreSQL JDBC Driver to remove the synchronised there (which is the JDBC driver I care most about). So yeah, I'm good to go with Virtual Threads. Just waiting on Helidon Nima now.

**Reply by jnehlmeier** [reply]
> Just waiting on Helidon Nima now

Sounds nice but wondering how jetty with virtual threads compare against it on pure HTTP performance.

**Reply by rob-bygrave** [reply]
> how jetty with virtual threads compare against it on pure HTTP performance.

Indeed. The implementation of Nima is really simple and uses VT per connection. Jetty 12 core now looks like its "async only" and to me that implies that with Jetty 12 there is always going to be the handoff from worker threads to Virtual Threads (unlike Jetty 11).

The Nima HTTP 1.0 & 1.1 implementation to me looks really nice and simple. The Helidon team have been benchmarking against Netty (Helidon 3.x is Netty based) and they are "All In" on Nima and Virtual Threads.

**Reply by rob-bygrave** [reply]
> how jetty with virtual threads compare against it on pure HTTP performance

Oh, just to say this was the way I did all my original Loom / Virtual Threads performance testing. Ran Jetty using platform threads vs loom with various "simulations of IO wait" using LockSupport.  That and using JDK HttpClient with Virtual Threads.  This is how I got very comfortable with the performance of Loom.

I have not update those tests and re-ran them for ages. I believe Jetty 12 makes that harder so if I re-run those tests I'll probably stick to Jetty 11. Those tests could also now run with actual IO wait like actual JDBC calls to a DB rather than simulation via LockSupport.

**Comment by jnehlmeier** [comment]
For the other points:

> Remove @Draftable

I never used it as I haven't used Ebean extensively so far but I generally liked the idea.

`@Draftable` seems similar to `@SoftDelete`, `@History` and `@ChangeLog` in the sense that it modifies queries and behavior within the ORM for the whole object graph including lazy loading. It is always pretty tough to replicate something like that outside an ORM because you do not have control over the queries itself in terms of lazy loading. That is why it is something that differentiates ebean from other solutions.

Would it be possibly to define extension points within Ebean so that such mechanisms can be implemented outside of ebean as an extension / plugin? Or is it maybe already possible but very complex to discover in the current API?


> Remove @Inheritance (Existing code would require to flatten model onto root bean, *could be painful)

I don't use it but this would be pure hell for migration from JPA.

> Remove @ElementCollection (promote use DB JSON columns going forward, *could be painful)

I use it quite a bit with `Enum` and also query the corresponding table element table for search (think of entities being tagged by defined tags as enum values or similar). Would also be migration hell. I guess you meant DBArray here as alternative instead of json. 

> Replace [SQL] `@formula`, with a [Logical property based] `@formula2` ?? *could be painful

No real opinion except that I always prefer to not reference DB table/columns in strings other than the ones in entities.

> Remove ElasticSearch support

No real opinion other than full text search is usually important. But of course if could be implemented externally as long as you can detect that data has been changed.

> Remove Entities as JSON support

I tried once to serialize ebean entities to json using GSON but that failed miserably because it uses reflection. So I discovered Ebean JSON serialization but somehow I wasn't really able to configure it properly. Documentation should somewhere state that a JSON serializer must call methods and not use field based reflection. At the end I created DTOs + mappers with exactly what I needed and used GSON to serialize the DTOs. I would have loved to just be able to add some "ignore this field and that field" configuration and then serialize the entity.

**Reply by rob-bygrave** [reply]
> @ElementCollection ... I guess you meant DBArray here as alternative instead of json

You'd use `@DbArray` with Postgres, YugabyteDB, CockroachDB, and H2 for _collections of scalar types_. You'd use JSON storage for the other databases and _collections of embeddables_.

**Reply by rob-bygrave** [reply]
> Remove @Inheritance 

Inheritance looks innocent enough but it has runtime costs and complexity costs. Ask the question, would you ever use inheritance if you were using SQL and JDBC?  Going back to conceptual Logical Modelling and Physical Modelling then yes inheritance comes from that area and is pretty well defined _BUT_ in 2023 no RDBMS that we care about has the concept and so using ORM `@Inheritance` adds an impedance mismatch on the application side - to some extent we are choosing to occasionally hit ourselves on the nose. At the core is that the PK is no longer enough most of the time and we instead need the PK AND discriminator ... and the complexity grows from there and touches pretty much all the internals of an ORM.

It will interesting to see how I feel with a diff that removes it wrt how much complexity that removes from the code base.

**Reply by rPraml** [reply]
Hmm. We are using @inheritance with the @docstore annotation when storing jsons. We do not use it anymore when storing db-beans. 

I'll try to provide a testcase/showcase tomorrw, what we do exactly and what we really need. Then we can discuss, if there is a way to model that in ebean or we have to switch to alternatives (e. g. Jackson subtypes) .

**Comment by JamesWilson3** [comment]
Please don't remove "@ElementCollection" as we are currently also using it for our Enum (similar to [jnehlmeier](https://github.com/jnehlmeier)'s comment above )

**Reply by rob-bygrave** [reply]
`@ElementCollection` has not been removed (from the 15x branch). 

To be clear, there is a ebean-15x branch that has all those other changes already made and that branch has been sitting there ready to go pretty much for a few weeks now. I've been keeping that branch up to date (rebasing).  There will be `14.x` releases that are pretty much the same as `13.x` but with jakarta and javax reversed + some deprecated methods removed + some improved default values like default jdbc persistBatch size increased (from the current default of 20 which imo is too low).

**Reply by JamesWilson3** [reply]
Thank you so much for the clarification and the work you've put into the ebean!

**Comment by gatno** [comment]
Sad for us to hear that @Inheritance is being removed. We have used it successfully many times

**Reply by rbygrave** [reply]
Ok. You can try and make the case for it? You should probably look to address the issues around runtime performance (like reference beans require a query) etc and perhaps. Ultimately its a cost/benefit call right and that has to also take into account the internal complexity than gets removed when the feature is removed (not sure if you've looked at the commits that removed the features).

**Reply by tbee** [reply]
We're stumbling onto (single table) inheritance being removed as well. Although I understand that inheritance often is overused, polymorphism **is** a core part of object oriented programming, and removing this will have a great impact on allowed architectures. There is nothing wrong with inheritance, if used correctly.

**Reply by VictorJimenezKwast** [reply]
We've also been happily using Inheritance in some of our entities for many years in a fairly large JEE application (+- million loc) and will be impacted a lot by this removal. 

Looking at the diff and the many properties, if statements and tests that were removed I understand the want to reduce complexity in Ebean. Inheritance is also not my preferred strategy to apply to most cases but it's part of the language and the JPA spec. Removing this will make it harder for other developers to make the move from Hibernate (or other JPA implementations) to Ebean.

I much prefer Ebean to Hibernate due to better query builders, explicit saves and the way it deals with loading relations. Still, I'm finding it harder to justify to my team to stay on Ebean.

**Reply by rbygrave** [reply]
So firstly, we do have some time to gather feedback and comments on `@Inheritance`. This means that we will continue to release in parallel 14.x, 14.x-javax, 15.x and these are all kept in sync in terms of bug fixes and feature changes. So 15.x is the same as 14.x but with those features removed including `@Inheritance`.

We are keeping these releases in parallel because generally speaking people are not reading release notes etc and so there is some expectation of some delay in terms of folks finally getting to update apps and only then coming to here at that point to give feedback - and that's ok but to some extent we can't wait forever for everyone to participate. The first 15.x release was Feb so that's already 7 months ago.

There is some possibility that inheritance makes it back into 15.x and we will probably know in Feb how it's going when 15.x would then have been out there for 1 year. We will have more comments added here, more participation, more feedback etc.



> nothing wrong with inheritance

When we have a relationship **_To_** an entity bean with inheritance, we can no longer instantiate a reference bean for that **_without_** the discriminator.  That is, the foreign key value by itself is no longer sufficient to create a reference bean and instead we additionally require the discriminator value and in order to get that ebean will add an extra sql join to that table [that has the inheritance] to obtain the discriminator value. We have this same scenario with `database.reference(Class<?>, idValue)` - this has to hit the database in the inheritance case [when normally this is just in-memory super cheap].

So this is the first extra cost we get when we use `@Inheritance` and in a strongly typed language like Java there really isn't any avoiding this issue unless we don't use `@Inheritance`.

Another thing we hit with single table inheritance is how we real with NOT NULL constraints on properties that are not on the base type. IMO there isn't really an optimal solution there. 

Internally within ebean we get complexity when we have an inheritance hierarchy with different relationships [ManyToOne, OneToMany etc] on different parts of the inheritance hierarchy. [We have "different shapes" with "different paths"]. 


## Alternatives to `@Inheritance`

An alternative approach to `@Inheritance` is for the ORM not to do it and instead deal with it in the application. For this, the inheritance is "flattened" on the ORM entity and then one approach is for the subtypes to be represented by interfaces that the now concrete entity class implements.

e.g. Vehicle with subtypes Car and Truck.  All the properties of Car and Truck are flattened onto Vehicle and Vehicle becomes a concrete type and the only `@Entity` and Car and Truck become interfaces that Vehicle implements. A method can be added that will return the appropriate interface based on the discriminator value [but this is now owned by the application code].

This approach is along the lines that we pull out the `@inheritance` functionality from the ORM into the application and now there is literally no "orm functionality" to hide how this works in terms of SQL and the database. There is nothing hidden from the developer here [e.g. when the devs make design decisions like non null or extra relationships the implications of these design decisions are now more obvious]. The upside is that any developer coming here will see how in works and its a pretty direct one to one mapping between the orm model and the SQL database model.

This also means that creating a reference bean for `Vehicle` is now "cheap" again with no discriminator value required and no extra join required.



For myself, keen to keep any and all feedback coming and see how this plays out. 

Remember that 14.x releases will continue and master branch is still 14.x and we are keeping that going for at least another 5 months [so all bug fixes and features still go into master branch which is 14.x and all changes there are merged back into 15.x branch].

**Reply by tbee** [reply]
I do not understand why single table inheritance would require an additional join. "Single table" says it already: there is only one table, what extra join can there be? All it should take is an in-memory map of discriminator values to actual class, and upon instantiation a lookup in that table which class to instantiate. I can see that partial selects are not an option anymore, because you first need to get the discriminator to decide what fields should be fetched, and with that select you might as well get the other fields as well. But that IMHO is a trade off for the developer to make. 

And yes, using @Inherited often has drawbacks; in the most optimal setup all the data is in a base class, and the extending classes only have differences on code level. Suppose Vehicle has a method calculateMaxCargo(), which differs only in code between Car and Truck. This is a natural fit for OO languages, where the use of inheritance is perfectly fine, and has no issues with any database constraints. Being forced to use interfaces and implementations is not a natural fit. 

But even if there are database complexities, like columns that are NOT NULL but not mapped in all extending classes, it is not the job of the ORM to deal with that. It is a design consideration of the developer; either use Inherited and somehow deal with that NOT NULL, or use another pattern. 

Nicely undermining the point I'm trying to make: one could argue that if you want inheritance in the domain model, trying to shoe-horn persistency onto the same model is not a good idea. JPA annotations are a way to re-use the domain model classes for persistency, which may not be at all suited for that. If the domain model classes cannot be easily used, a repository layer with DTO alike classes specifically for persistency (still using the annotations of course) may be needed. Maybe only for the one entity. When persisting a domain class, it first needs to be converted to repository layer classes, and vice versa. Not optimal, but IMHO better than not using natural OO architectures - if they are a good fit for a situation. 

But the point remains: there is always a price to pay, in code complexity, in less optimal database constraints, everything is a trade-off. But it should be up to the developer to decide where that price is due; not supporting inheritance in an OO language is IMHO a bad idea.

**Reply by rbygrave** [reply]
> why single table inheritance would require an additional join

Any table with a foreign key column to a table with inheritance.

e.g. Trip has a ManyToOne to Vehicle ... and Vehicle uses inheritance with Car and Truck subtypes.

The trip table as a fk column `vehicle_id` 

When we query trips, the orm reads the vehicle_id column, but that is not sufficient to create a reference bean for vehicle because it is using inheritance and it needs to create the reference bean with the correct type. It needs the discriminator value, and that is over in the vehicle table. So when we query trip it adds a join to vehicle such that it can read the value from the vehicle_id column AND the discriminator value. This is the case that means we get an extra join required purely due to `@Inheritance`.

**Comment by mschneider-lise** [comment]
We are considering switching from JPA to Ebean because of the following features:
- `@Draftable`
- `@History`
- `better query syntax`
- `better handling of lazy/eager loading`

I'm not sure if we will stick with Ebean if `@Draftable` is going to be removed. Unfortunately, having drafts is a requirement in our project and we really don't want to implement that ourselves. 😄 
Is there an update on removing/keeping `@Draftable`?

**Reply by rbygrave** [reply]
> update on removing/keeping @Draftable?

No real update really. The thing going against `@Draftable` is that as far as I know it is very very rarely used / it's a feature that hardly anyone uses or needs [as far as I am aware]. There is a chance that a whole bunch of existing users provide feedback and there is still time for that [because people can be relatively slow in upgrading apps etc] but my gut says we likely won't see that.

The plan is to continue to release in parallel both 14.x and 15.x and get explicit feedback. Without explicit feedback we can go off maven central download numbers for a rough picture of how many apps are stuck on 14.x vs got to 15.x [noting that will still be a group of apps that are in maintenance on earlier versions where we don't really know].

----

I should do an update on the priorities as I see them. It might also be worth noting that I'm also the creator of the avaje libraries like avaje-inject [https://avaje.io/inject/  - Dependency Injection as source code generation / a cross between Dagger2 and Spring DI]. These have matured nicely now but occasionally can take some "open source time" away from Ebean.

Ebean has about 4 things that are a high priority to get over the line with the first being truely read only immutable entity graphs so to some extent its about getting these things done and seeing where we are at.

**Reply by tbee** [reply]
Immutable entity graphs... If that can be done in a somewhat maintainable fashion. Interesting.

**Comment by tbee** [comment]
What is the status on these thoughts? Especially the Inheritance annotation?

**Reply by rbygrave** [reply]
No real change, still releasing the 14.x and 15.x branches in parallel etc. For myself, honestly no more thinking has gone on [busy on other things] so still running both releases and still looking for feedback.

There won't be a decision on Inheritance and the whole 14.x [and 16.x branch which is the follow-on from 14.x] for probably another year at this stage. That is, there is a bunch of work that needs to be done which I want to do which will happen before we get to circling back to Inheritance etc.
- The immutable graphs `query.setUnmodifiable(true)` is in Release candidates awaiting feedback
- Looking at a change to the load content (lazy loading batch by type rather than path)
- Release an open source version of the ebean insight server (collects all ebean metrics including query stats, plus the ability to dynamically collect query plans)
- Improved `@Formula` based on logical model rather than physical [columns etc]
- Plus other stuff like  maybe use avaje-jsonb internally


At this point maintaining the 15.x branch has gone well, meaning that dealing with merge conflicts hasn't been problematic. That is, as long as this isn't painful maintaining 2 branches we can continue with the 2 branches and continue to release 14.x and 15.x in parallel. [which will soon be 16.x and 17.x as we bump the major versions for the "Unmodifiable" feature] 


> Inheritance

At this stage I have only heard from 1 team about the migration away from Inheritance [and they actually did that prior to the 15.x branch]. I'm still keen to hear from others along the lines ... "We tried to migrate one entity from using Inheritance, we flattened the entity, ... what we found was ...".


In summary, there is still a lot of work ahead before we get back to Inheritance and still keen to get explicit feedback on it.

**Reply by tbee** [reply]
EBean is keeping you busy 😄 . We'll aim for the 16 branch then.
